### PR TITLE
React + Pyodide post

### PR DESCRIPTION
**Pull request by xhlulu**
Hi @rth, here's the draft. I have yet to create the JS fiddle since I thought you might want to host them under a `pyodide` account. Let me know your thoughts and feel free to make edits directly if needed!

closes https://github.com/pyodide/pyodide-blog/issues/11

Rendered version: https://deploy-preview-13--pyodide-blog.netlify.app/posts/react-in-python-with-pyodide/

*Estimated publication date*: 2022/1/19

### CODE REVIEW

**Review by rth — COMMENTED** [review]
Thanks for writing this @xhlulu!

It's a very cool demo of mixing React and Pyodide, and also a good illustration of using a JS library from Python using type conversions in general. @hoodmane would like it, I think.

I don't have experience with React, but overall I found it instructive and pleasant to read. Also I think it has a good balance between discussions, inline code examples, and links to more complex code parts.

A few minor comments below but overall it looks good to me.
>  I have yet to create the JS fiddle since I thought you might want to host them under a pyodide account. 

I created the Pyodide JSFiddle account, but in the end I think it might be easier to use the [GH integration](https://docs.jsfiddle.net/github-integration/untitled-1). As far as I understand we would need to create a subfolder per demo, rename files to the expected names, and then something like,
```
https://jsfiddle.net/gh/get/library/pure/xhlulu/pyodide-blog/tree/main/demos/react-in-pyodide/demo1/
```
should work. With a change of "xhlulu" -> "pyodide" before merging.

Also could you please provide a one-sentence bio? for the case when we manage to make https://github.com/pyodide/pyodide-blog/issues/14 work

**Comment by xhlulu** [comment]
@rth Sure! Here's my bio:

> Xing Han Lu is a M.Sc. student at Mila, where he works on NLP research, and previously worked as a ML developer at Plotly. He authored ML libraries on GitHub and ML notebooks on Kaggle, where he is a Notebooks Grandmaster.

Let me know if there's some place I can add that in the repo. Feel free to truncate. 

I'll work on the rest later today/tomorrow.

File: content/posts/react-in-pyodide.md [hunk 1]
```diff
@@ -0,0 +1,433 @@
+---
+title: "React in Pyodide"
+date: 2021-12-13T16:34:44-05:00
+draft: true
+tags: ["announcement"]
+author: "Xing Han Lu"
```
[the remark this answers is not in the document]
Comment (xhlulu) [reply content/posts/react-in-pyodide.md:6]: How do I add a link? I presume regular markdown doesn't work here?

**Comment by xhlulu** [comment]
Unfortunate i'm having trouble make the JS fiddle GH integration work. Not really sure what exactly i'd need to do. let me know if you have an idea!

**Comment by rth** [comment]
Thanks @xhlulu ! I'll look into the JS fiddle issue. 

What do you think about changing the title from,
```
React in Pyodide
```
to
```
React in Python with Pyodide
```
?

Otherwise LGTM. In terms of publication date, I was thinking around 2022/01/5 because there is another post tied to the release that we need to publish first: https://github.com/pyodide/pyodide-blog/wiki/Publication-schedule

**Comment by xhlulu** [comment]
Sounds good to me! I think the title change is okay!

**Comment by hoodmane** [comment]
This is really cool! Will try to do a detailed review in the next few days.

**Comment by rth** [comment]
OK, after some trial and error the jsfiddle integration works:
 - https://jsfiddle.net/gh/get/library/pure/rth/pyodide-blog/contents/demos/react-in-pyodide/demo-3/ 
 
(and changing the demo number from 1 to 5). Currently, this is hosted on my fork but I updated this PR so once it's merged and we rename the main branch, it should work on pyodide/pyodide-blog. 

About jsfiddle their docs are not good and it doesn't seem to be actively maintained, however as far as I can tell after quickly looking that's the only solution that allows an easy sync from Github. Also to keep in mind,
 - one has to split the example between `demo.html` and `demo.js` which is maybe a bit less intuitive. Also, the libraries to load go to the HTML file without specifying the `head` tag, which is not great but that's the only way to make it work apparently.
 - there is some caching of a few minutes. So if you push a demo file, check it in jsfiddle then change the file again and check again you will likely see the previous version. The solution is either wait a bit or make a new unique folder for the demo.

So from my side, it looks ready. Just a last few points,  @xhlulu  would you mind:
 - checking the title/description in each demo.description files (and updating it if needed)
 - add the links to the demos using the `https://jsfiddle.net/gh/get/library/pure/pyodide/pyodide-blog/contents/demos/react-in-pyodide/demo-3/ ` URL
 -  > Let me know if there's some place I can add that in the repo
    
    Maybe for now let's add a section "About the author" at the end,  in italic or some different style or font than the other sections?
 -  Could we update the demos to Pyodide 0.19.0 ? I haven't tried but I guess it should work without changes.

**Comment by rth** [comment]
Also if you have some preference on how this post should be announced on Twitter let me know, otherwise, I'll come up with something.

**Comment by xhlulu** [comment]
> Also if you have some preference on how this post should be announced on Twitter (under the Pyodide account) let me know, otherwise, I'll come up with something.

Sure that's good! I can retweet the post.

> OK, after some trial and error the jsfiddle integration works

Awesome! Thank you @rth for the amazing detective 🕵️ work here :) Glad you were able to get it working in the end.

> checking the title/description in each demo.description files (and updating it if needed)

Just did that and they all looked good.

> add the links to the demos

Done

> Maybe for now let's add a section "About the author" at the end, in italic 

Just added that.

> Could we update the demos to Pyodide 0.19.0

I just tried and it gives a weird error: https://jsfiddle.net/1dc98aeb/


```
<a class='gotoLine' href='#[object Error] { ... }'>[object Error] { ... }</a>
```

**Comment by rth** [comment]
Great, thanks @xhlulu !

> I just tried and it gives a weird error

If you open the console the actual traceback is I think,
```
  File "/lib/python3.9/site-packages/_pyodide/_base.py", line 345, in run_async
    coroutine = eval(self.code, globals, locals)
  File "<exec>", line 82, in <module>
  File "???.js", line -1, in <javascript frames>
JsException: Error: This borrowed proxy was automatically destroyed at the end of a function call. Try using create_proxy or create_once_callable.
```
which is due to the change in https://github.com/pyodide/pyodide/pull/1573. I haven't looked at your code in detail, but I think your examples are a good use case for that feature, so it would be nice to look into a fix. 
Also more specific error message indicating which proxy or LoC  is in question, wouldn't hurt, otherwise it's hard to debug when running large code snippets cc @hoodmane

File: content/posts/react-in-python-with-pyodide.md [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+
+
+> **About this author:** *Xing Han Lu is a graduate student at Mila, where he works on NLP research, and previously worked as a ML developer at Plotly. He authored ML libraries on GitHub and publish code and data on Kaggle, where he is a Code Grandmaster.*
```
Comment (rth) [inline content/posts/react-in-python-with-pyodide.md:435]: Thanks the rendering looks good! Do you want to also link to your personal website and or mention your twitter username?

```suggestion
> **About the author** 
>
> *[Xing Han Lu](https://xinghanlu.com/) ([@xhluu](https://twitter.com/xhluu)) is a graduate student at Mila, where he works on NLP research, and previously worked as a ML developer at Plotly. He authored ML libraries on GitHub and publish code and data on Kaggle, where he is a Code Grandmaster.*
```

**Comment by rth** [comment]
Merging the current version. I opened https://github.com/pyodide/pyodide-blog/issues/20 to track updating the Pyodide version in examples. 

Thanks again @xhlulu !

**Comment by xhlulu** [comment]
Thank you @rth for the great work on editing and reviewing the post :)

**Comment by hoodmane** [comment]
>   `File "???.js", line -1, in <javascript frames>`
> ...
> Also more specific error message indicating which proxy or LoC is in question

Javascript error tracebacks could be improved a lot by making a better effort [here](https://github.com/pyodide/pyodide/blob/de61d5ab1f202ef6bcc522cf924e771dca726537/src/core/error_handling.c#L267):
```
    // TODO? Use stacktracejs to add more detailed info here.
    __PyTraceback_Add(HEAPU32[_error__js_funcname_string / 4],
                      HEAPU32[_error__js_filename_string / 4],
                      -1);
```
Figuring out what to put though is a bit confusing. I made a few attempts at this last spring, but it a bunch of different subtle questions about what should be done so I never quite got to it. I can make a PR that just does *something* and then we can discuss.

> indicating which proxy

Perhaps before we destroy it we can store the type and/or repr for use in errors?

**Pull request merged**